source/common/router/rds_impl.cc

```diff
@@ -159,21 +179,117 @@ Router::ConfigConstSharedPtr RdsRouteConfigProviderImpl::config() {
+  }
+
+  initialize_vhosts(route_config_proto_);
+
+  subscription_ = Envoy::Config::SubscriptionFactory::subscriptionFromConfigSource(
+          route_configuration.vhds().config_source(), factory_context.localInfo(),
+          factory_context.dispatcher(), factory_context.clusterManager(), factory_context.random(), *scope_,
+          "none",
+          "envoy.api.v2.VirtualHostDiscoveryService.DeltaVirtualHosts",
+          Grpc::Common::typeUrl(envoy::api::v2::route::VirtualHost().GetDescriptor()->full_name()),
+          factory_context.api());
+}
+
+void VhdsSubscription::onConfigUpdateFailed(const EnvoyException*) {
+  // We need to allow server startup to continue, even if we have a bad
+  // config.
+  init_target_.ready();
 }
 
+void VhdsSubscription::onConfigUpdate(const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+                                                const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+                                                const std::string& version_info) {
+  last_updated_ = time_source_.systemTime();
+  remove_vhosts(virtual_hosts_, removed_resources);
+  update_vhosts(virtual_hosts_, added_resources);
+  rebuild_route_config(virtual_hosts_, route_config_proto_);
+
+  const uint64_t new_hash = MessageUtil::hash(route_config_proto_);
+  if (!config_info_ || new_hash != config_info_.value().last_config_hash_) {
+    config_info_ = {new_hash, version_info};
+    stats_.config_reload_.inc();
+    ENVOY_LOG(debug, "vhds: loading new configuration: config_name={} hash={}", route_config_name_,
+              new_hash);
+    for (auto* provider : rds_subscription_->route_config_providers()) {
+      provider->onConfigUpdate();
+    }
+  }
+
+  init_target_.ready();
+}
+
+void VhdsSubscription::initialize_vhosts(const envoy::api::v2::RouteConfiguration& route_configuration) {
+  if (route_configuration.virtual_hosts_size() <= 0) { return; }
+  for(const auto& vhost : route_configuration.virtual_hosts()) {
+    virtual_hosts_.emplace(vhost.name(), vhost);
+  }
+}
+
+void VhdsSubscription::remove_vhosts(
+        std::unordered_map<std::string, envoy::api::v2::route::VirtualHost>& vhosts,
+        const Protobuf::RepeatedPtrField<std::string>& removed_vhost_names) {
+  for(const auto vhost_name : removed_vhost_names) {
+    vhosts.erase(vhost_name);
+  }
+}
+
+void VhdsSubscription::update_vhosts(
+        std::unordered_map<std::string, envoy::api::v2::route::VirtualHost>& vhosts,
+        const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources) {
+  // TODO validate aaded_resources
+  for(const auto& resource : added_resources) {
+    envoy::api::v2::route::VirtualHost vhost =
+            MessageUtil::anyConvert<envoy::api::v2::route::VirtualHost>(resource.resource());
+    vhosts.emplace(vhost.name(), vhost);
+  }
+}
+
+void VhdsSubscription::rebuild_route_config(
```

Comment:
Thanks for this contribution, this will be foundational to VHDS.

I'm going to wait until the API PR is merged before diving into this in depth, but as a starting point, can you please make sure things like formatting and naming match Envoy convention and style guide https://github.com/envoyproxy/envoy/blob/master/STYLE.md. E.g. this should be `rebuildRouteConfig`, below should be `for (const auto` etc. Running `fix_format` will help with a bunch.

Don't want to be too nit picky, but it will make the review easier to focus on the substance if you can clean up the style issues in advance.